### PR TITLE
fix: skip setting tcp_nodelay for unix domain sockets

### LINKED ISSUE #2030: Fix tcp_nodelay option on unix sockets
From discord:

>after updating to v1.11.0, dragonfly is refusing connections over unix sockets
errors in logs are attached

[message.txt](https://github.com/dragonflydb/dragonfly/files/12927108/message.txt)

### PR DESCRIPTION
**Pull request by dranikpg**
Fixes #2030

### CODE REVIEW

**Comment by kostasrim** [comment]
@dranikpg is this ready for review?

File: src/facade/dragonfly_connection.cc [hunk 1]
```diff
@@ -337,8 +337,17 @@ void Connection::HandleRequests() {
...
-    int val = 1;
-    CHECK_EQ(0, setsockopt(socket_->native_handle(), IPPROTO_TCP, TCP_NODELAY, &val, sizeof(val)));
+    int val = 0;
+
+    // Get the underlying address family first to avoid setting tcp options on unix domain sockets
+    socklen_t bufsize = sizeof(val);
+    int res = getsockopt(socket_->native_handle(), SOL_SOCKET, SO_DOMAIN, &val, &bufsize);
+    LOG_IF(ERROR, res != 0) << "Bad getsockopt call " << res;
+
+    if (res == 0 && (val == AF_INET || val == AF_INET6)) {
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc:347]: You'd ask, why not check with `socket_->IsUDS()`... Well its always false 😆 `UringSocket::Accept` just doesn't propagate this flag in the file-descriptor meta bits

You'd ask, why not just propagate this flag... Well, we `DCHECK(fd & UDS == 0)` in some places, like `RemoteEndpoint()` which is called for all connections 

So we have two bugs, but they cancel out each other. Not sure if I should fix both or just keep this fix

**Comment by dranikpg** [comment]
> @dranikpg is this ready for review?

I have a philosophical question (whether to fix two bugs which cancel out) so someone better first respond on that

**Comment by kostasrim** [comment]
> > @dranikpg is this ready for review?
> 
> I have a philosophical question (whether to fix two bugs which cancel out) so someone better first respond on that

To be or not to be? :thought_balloon:

File: src/facade/dragonfly_connection.cc (hunk 1 above)
Comment (dranikpg) [reply src/facade/dragonfly_connection.cc:347]: Yes... I'll fix it in this pr, so lets keep it as a draft for now

File: src/facade/dragonfly_connection.cc [hunk 2]
```diff
@@ -1096,16 +1091,34 @@ void Connection::EnsureAsyncMemoryBudget() {
...
+  return absl::StrCat(re.address().to_string(), ":", re.port());
 }
 
 std::string Connection::RemoteEndpointAddress() const {
+  if (socket_->IsUDS())
+    return "unix-domain-socket";
+
   auto re = socket_->RemoteEndpoint();
   return re.address().to_string();
 }
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc:1124]: I don't like this...What should I return? Technically is it well possible to fetch the endpoints uds file path with the same `getpeername` from `RemoteEndpoint` below. However it requires changing the return type everywhere from `boost::asio::ip::endpoint` to smth else that can hold generic strings

Comment (romange) [reply src/facade/dragonfly_connection.cc:1124]: I prefer this or "uds", not worth the time to improve it.

**Review by romange — APPROVED** [review]

**Pull request merged**